app/celery/process_ses_receipts_tasks.py

```diff
@@ -1,80 +1,166 @@
+import enum
+import traceback
 from datetime import datetime, timedelta
+from json import decoder
 
 import iso8601
+import requests
 from celery.exceptions import Retry
-from flask import current_app, json
+from flask import Blueprint, current_app, json, jsonify, request
 from sqlalchemy.orm.exc import NoResultFound
 
-from app import notify_celery, statsd_client
-from app.clients.email.aws_ses import get_aws_responses
+from app import notify_celery, redis_store, statsd_client
+from app.celery.validate_sns import validate_sns_message
 from app.config import QueueNames
 from app.dao import notifications_dao
+from app.errors import InvalidRequest, register_errors
 from app.models import NOTIFICATION_PENDING, NOTIFICATION_SENDING
 from app.notifications.notifications_ses_callback import (
     _check_and_queue_complaint_callback_task,
     check_and_queue_callback_task,
     determine_notification_bounce_type,
+    get_aws_responses,
     handle_complaint,
 )
 
+ses_callback_blueprint = Blueprint('notifications_ses_callback', __name__)
+DEFAULT_MAX_AGE = timedelta(days=10000)
+
+register_errors(ses_callback_blueprint)
+class SNSMessageType(enum.Enum):
+    SubscriptionConfirmation = 'SubscriptionConfirmation'
+    Notification = 'Notification'
+    UnsubscribeConfirmation = 'UnsubscribeConfirmation'
+
+
+class InvalidMessageTypeException(Exception):
+    pass
+
+
+def verify_message_type(message_type: str):
+    try:
+        SNSMessageType(message_type)
+    except ValueError:
+        raise InvalidMessageTypeException(f'{message_type} is not a valid message type.')
+
+
+# 400 counts as a permanent failure so SNS will not retry.
+# 500 counts as a failed delivery attempt so SNS will retry.
+# See https://docs.aws.amazon.com/sns/latest/dg/DeliveryPolicies.html#DeliveryPolicies
+# This should not be here, it used to be in notifications/notifications_ses_callback. It then
+# got refactored into a task, which is fine, but it created a circular dependency. Will need
+# to investigate why GDS extracted this into a lambda
+@ses_callback_blueprint.route('/notifications/email/ses', methods=['POST'])
+def sns_callback_handler():
+    message_type = request.headers.get('x-amz-sns-message-type')
+    try:
+        verify_message_type(message_type)
+    except InvalidMessageTypeException:
+        current_app.logger.exception(f"Response headers: {request.headers}\nResponse data: {request.data}")
+        raise InvalidRequest("SES-SNS callback failed: invalid message type", 400)
+
+    try:
+        message = json.loads(request.data.decode('utf-8'))
+    except decoder.JSONDecodeError:
+        current_app.logger.exception(f"Response headers: {request.headers}\nResponse data: {request.data}")
+        raise InvalidRequest("SES-SNS callback failed: invalid JSON given", 400)
+
+    try:
+        validate_sns_message(message)
```

Comment:
validate_sns_message() replaces the validatesns lib, imported from GC Notify, which seems to be no longer maintained at this point.

I ran into several issues with it and found https://github.com/boto/boto3/issues/2508 which illuminated all of them.

I had to replace M2crypto in the above example with oscrypto because the linux base image (Ubuntu Bionic 18.04) used by CG does not have a backport for the current version of M2crypto. Oscrypto is widely-used and well-maintained.

Reply:
FYI, if we _really_ needed to, we could validate sns certs using only openSSL

Reply:
function has been renamed validate_sns_cert and now lives in app.notifications.sns_cert_validator, but the above comments still apply